rfcs/2021-11-12-9811-vrl-vm.md

```diff
@@ -0,0 +1,164 @@
+# RFC 9811 - 2021-11-12 - VRL bytecode VM
+
+This RFC proposes implementing a bytecode VM. VRL will be compiled to the bytecode
+and executed by this VM, with the aim to significantly improve the performance of
+executing VRL scripts.
+
+## Context
+
+This RFC is a follow on from
+[rfcs/2021-10-14-9811-vrl-performance.md](https://github.com/vectordotdev/vector/pull/9812)
+
+## Scope
+
+### In scope
+
+This RFC is purely about developing a bytecode VM for VRL.
+
+### Out of scope
+
+Any other performance issues relating to VRL will be discussed in
+[rfcs/2021-10-14-9811-vrl-performance.md](https://github.com/vectordotdev/vector/pull/9812)
+
+## Proposal
+
+### Implementation
+
+Vrl compiles to an AST that is then walked during resolution. Each node in that
+tree is boxed and stored in disparate regions of memory. As a result walking
+the tree means that the CPU caches must be constantly swapped.
+
+Instead we can create a bytecode VM to store the execution of the Vrl program.
+
+Bytecode is essentially a big enum of instructions (that are represented by an integer
+using the [num-derive](https://crates.io/crates/num-derive),
+[num-traits](https://crates.io/crates/num-traits) crates):
+
+```rust
+#[derive(FromPrimitive, ToPrimitive, Copy, Clone, Debug, PartialEq, Eq)]
+pub enum OpCode {
+    Return = 255,
+    Constant,
+    Negate,
+    Add,
+    Subtract,
+    Multiply,
+    Divide,
+    Print,
+    Not,
+    Greater,
+    GreaterEqual,
+    Less,
+    LessEqual,
+    NotEqual,
+    Equal,
+    Pop,
+    JumpIfFalse,
+    Jump,
+    SetPath,
+    GetPath,
+    Call,
+    ...
+}
+```
+
+The Vm is a struct comprising of the following fields:
+
+```rust
+#[derive(Clone, Debug, Default)]
+pub struct Vm {
+    instructions: Vec<usize>,
+    values: Vec<Literal>,
+    targets: Vec<Variable>,
+    stack: Vec<Value>,
+    ip: usize,
```

Comment:
I know this is just an example for this RFC, but I strongly recommend using full names instead of acronyms, since not many people might/will be familiar with the terminology used.